File: course-schedule.cc
// Course Schedule
#define FOR(i, a, b) for (decltype(b) i = (a); i < (b); i++)
#define REP(i, n) FOR(i, 0, n)

class Solution {
public:
  bool canFinish(int n, vector<vector<int>> &a) {
    vector<int> d(n, 0);
    vector<vector<int>> e(n);
    int top = -1, num = 0;
    for (auto x: a) {
      d[x[0]]++;
      e[x[1]].push_back(x[0]);
    }
    REP(i, n)
      if (! d[i])
        d[i] = top, top = i;
    while (top != -1) {
      int x = top;
      top = d[top];
      num++;
      for (auto y: e[x])
        if (! --d[y])
          d[y] = top, top = y;
    }
    return num == n;
  }
};

File: course-schedule.nim
type cintArray = ptr UncheckedArray[cint]

proc canFinish(n: int, prerequisites: ptr UncheckedArray[cintArray], m: int, unused: pointer): bool {.exportc.} =
  var
    d = newSeq[int](n)
    e = newSeq[seq[int]](n)
    top = -1
    num = 0
  for x in toOpenArray(prerequisites, 0, m-1):
    inc d[x[0]]
    e[x[1]].add x[0]
  for i in 0..<n:
    if d[i] == 0:
      d[i] = top
      top = i
  while top >= 0:
    let x = top
    top = d[top]
    inc num
    for y in e[x]:
      dec d[y]
      if d[y] == 0:
        d[y] = top
        top = y
  return num == n
